allennlp/models/bidaf.py

```diff
@@ -0,0 +1,348 @@
+
+        # Shape: (batch_size, modeling_dim)
+        span_start_representation = weighted_sum(modeled_passage, span_start_probs)
+        # Shape: (batch_size, passage_length, modeling_dim)
+        tiled_start_representation = span_start_representation.unsqueeze(1).expand(batch_size,
+                                                                                   passage_length,
+                                                                                   modeling_dim)
+
+        # Shape: (batch_size, passage_length, encoding_dim * 4 + modeling_dim * 3)
+        span_end_representation = torch.cat([final_merged_passage,
+                                             modeled_passage,
+                                             tiled_start_representation,
+                                             modeled_passage * tiled_start_representation],
+                                            dim=-1)
+        # Shape: (batch_size, passage_length, encoding_dim)
+        encoded_span_end = self._span_end_encoder(span_end_representation)
+        # Shape: (batch_size, passage_length, encoding_dim * 4 + span_end_encoding_dim)
+        span_end_input = torch.cat([final_merged_passage, encoded_span_end], dim=-1)
+        span_end_logits = self._span_end_predictor(span_end_input).squeeze(-1)
+        span_end_probs = masked_softmax(span_end_logits, passage_mask)
+
+        output_dict = {"span_start_logits": span_start_logits,
+                       "span_start_probs": span_start_probs,
+                       "span_end_logits": span_end_logits,
+                       "span_end_probs": span_end_probs}
+        if span_start:
+            # Negative log likelihood criterion takes integer labels, not one hot.
+            if span_start.dim() == 2:
+                _, span_start = span_start.max(-1)
+            loss = self._loss(span_start_logits, span_start.view(-1))
+            if span_end.dim() == 2:
+                _, span_end = span_end.max(-1)
+            loss += self._loss(span_end_logits, span_end.view(-1))
+            output_dict["loss"] = loss
+
+        return output_dict
+
+    def predict_span(self, question: TextField, passage: TextField) -> Dict[str, Any]:
+        """
+        Given a question and a passage, predicts the span in the passage that answers the question.
+
+        Parameters
+        ----------
+        question : ``TextField``
+        passage : ``TextField``
+
+        Returns
+        -------
+        A Dict containing:
+
+        span_start_probs : numpy.ndarray
+        span_end_probs : numpy.ndarray
+        best_span : (int, int)
+        """
+        question.index(self._vocab)
+        passage.index(self._vocab)
+        # TODO(mattg): we should make the lengths an optional parameter to Instance.as_array()
+        question_lengths = question.get_padding_lengths()
+        question_input = arrays_to_variables(question.as_array(question_lengths))
+        for input_array in question_input.values():
```

Comment:
Idea: We could add this functionality as a `assert_batch_dimension` flag to `arrays_as_variables`.